peps/pep-0790.rst

```diff
@@ -0,0 +1,400 @@
+  a list, with each traceback composited from any intermediate groups.
+
+- ``with BaseException.preserve_context() [as exc]:``, a context manager which
+  saves and restores the ``__context__`` and ``__suppress_context__``
+  attributes, to raise an exception unchanged within an ``except:`` block.
+
+We expect this to enable more concise expression of error handling logic in
+many medium-complexity cases.  Without them, exception-group handlers will
+continue to discard intermediate tracebacks and mis-handle ``__context__``
+exceptions, to the detriment of anyone debugging async code.
+
+
+Motivation
+==========
+
+As exception groups come into widespread use, library authors and end users
+often write code to process or respond to individual leaf exceptions, for
+example when implementing middleware, error logging, or response handlers in
+a web framework.
+
+`Searching GitHub <https://github.com/search?q=%2Ffor+%5Cw%2B+in+%5Beg%5D%5Cw*%5C.exceptions%3A%2F+language%3APython&type=code>`__
+found four implementations of ``flat_exceptions()`` by various names in the
+first sixty hits, of which none handle tracebacks.[#numbers]_  The same search
+found thirteen cases where ``.flat_exceptions()`` could be used.  We therefore
+believe that providing an upstream method with proper traceback preservation
+will improve error-handling and debugging experiences across the ecosystem.
+
+
+The rise of exception groups has also made re-raising exceptions caught by an
+earlier handler much more common: for example, web-server middleware might
+unwrap ``HTTPException`` if that is the sole leaf of a group:
+
+.. code-block:: python
+
+    except* HTTPException as group:
+        first, *rest = group.flat_exceptions()  # get the whole traceback :-)
+        if not rest:
+            raise first
+        raise
+
+However, this innocent-seeming code has a problem: ``raise first`` will do
+``first.__context__ = group`` as a side effect.  This discards the original
+context of the error, which may contain crucial information to understand why
+the exception was raised.  In many production apps it also causes tracebacks
+to balloon from hundreds of lines, to tens or even `hundreds of thousands of
+lines <https://github.com/python-trio/trio/issues/2001#issuecomment-931928509>`__ -
+a volume which makes understanding errors far more difficult than it should be.
+
+A new ``BaseException.preserve_context()`` method would be a discoverable,
+readable, and easy-to-use solution for these cases.
+
+
+Specification
+=============
+
+Adding ``BaseExceptionGroup.flat_exceptions()``
+-----------------------------------------------
+
+A new method ``flat_exceptions()`` will be added to ``BaseExceptionGroup``, with the
+following signature:
+
+.. code-block:: python
+
+    def flat_exceptions(self) -> list[BaseException]:
+        """
+        Return a flat list of all 'leaf' exceptions.
+
+        If the exception is not a ``BaseExceptionGroup``, return ``[eg]``.
```

Comment:
```suggestion
        If the exception is not a ``BaseExceptionGroup``, return ``[self]``.
```
"eg" is not defined. I assume it stands for ExceptionGroup, but we've just established that it isn't one :)

Reply:
In fact this docstring is just out of date; since this method is defined on `BaseExceptionGroup` (not `BaseException`), `self` _must_ be a group.  Will update.